### PR TITLE
CRM-19767 - Add case tokens to email activities

### PR DESCRIPTION
**Pull request by colemanw**
In CiviCRM core, case tokens had only been partially implemented. 

- When doing _Print/Merge Document_ from a case, tokens for `{case.id}`, etc, _were_ available.
- When doing _Send Email_ from a case, the tokens were missing.

This fixes that.

* [CRM-19767: Add Case Tokens to Case Activities](https://issues.civicrm.org/jira/browse/CRM-19767)

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
How's about a test :-)

**Comment by colemanw** [comment]
Looks like those bloated functions are untestable as-is. Would take some 
refactoring. I'll see what I can do.

**Comment by eileenmcnaughton** [comment]
CRM_Contribute_Form_AdditionalPaymentTest or CRM_Contribute_Form_Task_InvoiceTest might provide a starting point

**Comment by eileenmcnaughton** [comment]
test this please

File: CRM/Utils/Token.php [hunk 1]
```diff
@@ -1576,7 +1576,7 @@ public static function replaceEntityTokens($entity, $entityArray, $str, $knownTo
 
   /**
    * @param int $caseId
-   * @param int $str
+   * @param string $str
    * @param array $knownTokens
    * @param bool $escapeSmarty
```
Comment (tunbola) [inline CRM/Utils/Token.php]: There should be space here before the `@return string`

File: CRM/Contact/Form/Task/PDFLetterCommon.php [hunk 2]
```diff
@@ -352,6 +352,8 @@ public static function processMessageTemplate($formValues) {
    * Process the form after the input has been submitted and validated.
    *
    * @param CRM_Core_Form $form
```
Comment (tunbola) [inline CRM/Contact/Form/Task/PDFLetterCommon.php]: There should be a space here also before `@throws \CRM_Core_Exception`

**Review by tunbola — CHANGES_REQUESTED** [review]

**Comment by tunbola** [comment]
Also in addition to what was already said, would be great to have some tests for this change. :)

File: CRM/Contact/Form/Task/PDFLetterCommon.php (hunk 2 above)
Comment (colemanw) [inline CRM/Contact/Form/Task/PDFLetterCommon.php]: Done

File: CRM/Utils/Token.php (hunk 1 above)
Comment (colemanw) [inline CRM/Utils/Token.php]: Done

**Comment by colemanw** [comment]
@guanhuan will you assign adding the test please?

**Comment by eileenmcnaughton** [comment]
I'm going to close this because it needs rebasing & I think it's better to re-open once (if) rebased than to keep it open

**Comment by colemanw** [comment]
Drat. Github won't allow me to reopen a PR after rebasing it because of the force push it considers the branch to be different :(.  I've opened a new one instead.

**Pull request closed without merging**